### PR TITLE
feat: rbac overrides in package.json

### LINKED ISSUE #814: Have a way to create overrides for RBAC in a source file.
### Is your feature request related to a problem? Please describe.

We want to leverage `package.json` to defined RBAC settings that will work with RBAC and create an excellent example test. 

### Describe the solution you'd like

- **Given** we want to create a reproducible way to generate scoped RBAC 
- **When** we define the RBAC in package.json and run a build
- **Then** we have our customized RBAC

Definition of done: 
- [ ] bring rbac rules to values.yaml

### PR DESCRIPTION
**Pull request by schaeferka**
## Description
We want to leverage package.json to defined RBAC settings that will work with RBAC and create an excellent example test.

Describe the solution you'd like
Given we want to create a reproducible way to generate scoped RBAC
When we define the RBAC in package.json and run a build
Then we have our customized RBAC
...

End to End Test:    
(See [Pepr Excellent Examples](https://github.com/defenseunicorns/pepr-excellent-examples))

## Related Issue

Fixes #814 

Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging
- [ ] Unit, [Journey](https://github.com/defenseunicorns/pepr/tree/main/journey), [E2E Tests](https://github.com/defenseunicorns/pepr-excellent-examples), [docs](https://github.com/defenseunicorns/pepr/tree/main/docs), [adr](https://github.com/defenseunicorns/pepr/tree/main/adr) added or updated as needed
- [ ] [Contributor Guide Steps](https://docs.pepr.dev/main/contribute/#submitting-a-pull-request) followed

### CODE REVIEW

File: src/lib/assets/helm.ts
Comment (cmwylie19) [inline src/lib/assets/helm.ts]: can you use upstream types, with kubernetes/client-node? 

These are the same type:

```plaintext
$ k explain clusterRole.rules
GROUP:      rbac.authorization.k8s.io
KIND:       ClusterRole
VERSION:    v1

FIELD: rules <[]PolicyRule>

$ k explain role.rules 
GROUP:      rbac.authorization.k8s.io
KIND:       Role
VERSION:    v1

FIELD: rules <[]PolicyRule>


```

Comment (schaeferka) [reply src/lib/assets/helm.ts]: the interfaces are now gone

File: journey/pepr-build.ts
Comment (cmwylie19) [inline journey/pepr-build.ts]:
```suggestion
```

File: docs/030_user-guide/120_customization.md
Comment (cmwylie19) [inline docs/030_user-guide/120_customization.md]: this should be role and clusterRole since we only have one of each. The plural is misleading

File: src/lib/assets/yaml.ts
Comment (cmwylie19) [inline src/lib/assets/yaml.ts]: what is the reason for this change?

File: src/lib/assets/yaml.ts
Comment (cmwylie19) [inline src/lib/assets/yaml.ts]: we need to use the RBAC that we create

File: src/lib/assets/yaml.ts
Comment (cmwylie19) [inline src/lib/assets/yaml.ts]: we need to use the rbac that we already create

File: src/lib/assets/yaml.ts
Comment (schaeferka) [reply src/lib/assets/yaml.ts]: Reverted to original

File: src/lib/assets/yaml.ts
Comment (schaeferka) [reply src/lib/assets/yaml.ts]: [@Case](https://defense-unicorns.slack.com/team/U05AJDH3ME0) Quick question for you on the RBAC PR. For this section of code in yaml.ts:

```yaml
 rbac: {
      clusterRoles: customClusterRoleRules.map((rule, index) => ({
        name: `custom-cluster-role-${index + 1}`,
        rules: rule,
      })),
      roles: customStoreRoleRules.map((rule, index) => ({
        name: `custom-role-${index + 1}`,
        rules: rule,
      })),
```

You commented that we need to use the RBAC we create. Do you mean the scoped/admin rules that are always generated, the custom rules from the package.json file, or both? I just want to make sure since the original yaml.ts file didn't include the scoped/admin rules in the overrides file.

File: src/lib/assets/yaml.ts
Comment (cmwylie19) [reply src/lib/assets/yaml.ts]: use singular role and clusterRole

Comment (schaeferka) [reply src/lib/assets/yaml.ts]: Changed to singular

File: docs/030_user-guide/120_customization.md
Comment (cmwylie19) [inline docs/030_user-guide/120_customization.md]:
```suggestion
| `rbac`          | Custom RBAC rules                    | `"<role/clusterRole>": {"rules": [{"apiGroups": ["<apiGroups>"], "resources": ["<resources>"], "verbs": ["<verbs>"]}]}` |
```

File: src/lib/module.ts
Comment (cmwylie19) [inline src/lib/module.ts]: Upstream types `PolicyRule`

File: src/lib/types.ts
Comment (cmwylie19) [inline src/lib/types.ts]: shouldn't this be RBAC and maybe an array of PolicyRules?

Comment (cmwylie19) [reply src/lib/types.ts]: Yeah i think it actually should since there can be more than one rule set

Comment (schaeferka) [reply src/lib/types.ts]: Updated to:

```ts
export interface CapabilityExport extends CapabilityCfg {
  bindings: Binding[];
  hasSchedule: boolean;
  rbac?: PolicyRule[];
}
```

File: src/lib/module.ts
Comment (schaeferka) [reply src/lib/module.ts]: Removed the Rule interface and updated to import and use PolicyRule.

File: src/lib/assets/index.ts
Comment (cmwylie19) [inline src/lib/assets/index.ts]:
```suggestion
```

Comment (cmwylie19) [reply src/lib/assets/index.ts]: remove comment

File: src/lib/assets/rbac.ts
Comment (cmwylie19) [inline src/lib/assets/rbac.ts]: not sure we need to read this in again, we have access to this in the function `allYaml` that calls this function, it is the assets, is there a reason for this?

Comment (schaeferka) [reply src/lib/assets/rbac.ts]: The `readRBACFromPackageJSON` function is used to get the custom RBAC rules from the `package.json`. It is called from the `clusterRole` function in `rbac.ts`, which then merges the custom rules into the generated RBAC rules. 

Currently, we are calling the `clusterRole` function twice - once in the `overridesFile` function and once in the `allYaml` function. The `assets/index.ts` file calls `overridesFile` and `allYaml` separately. But, as part of the `yaml.ts` refactoring we could improve this so that `clusterRole` is only called once. For example, in `assets/index.ts` we could call `clusterRole` once to get the merged list of rbac rules, and then hand that list to both `overrides` and `allYaml`.

File: src/lib/module.ts
Comment (cmwylie19) [inline src/lib/module.ts:42]: This is out of scope for this issue, better scoped for a new issue

Comment (cmwylie19) [reply src/lib/module.ts:42]: Going to go ahead and prioritize issue https://github.com/defenseunicorns/pepr/issues/1327 so we can get this merged.

**Review by cmwylie19 — DISMISSED** [review]

File: src/lib/assets/rbac.ts
Comment (cmwylie19) [reply src/lib/assets/rbac.ts]: glad to see we are only reading the module config once again.

File: journey/pepr-build-wasm.ts
Comment (samayer12) [inline journey/pepr-build-wasm.ts:17]: This test doesn't actually assert anything, it just performs a `mkdir` operation without an `expect()` statement to ensure success. We should fix that. We should also perform test cleanup (i.e., removing the directory) after the test finishes.

Not sure if this will cause issues with the follow-on tests, since what we have in `journey/` usually aren't isolated from each other. Possibly out-of-scope for this issue and deserves a follow-on tech debt chore.

File: src/lib/assets/deploy.ts
Comment (samayer12) [inline src/lib/assets/deploy.ts]: Why do we destructure the `assets.config` object here instead of passing in `assets.config`?

**Review by samayer12 — CHANGES_REQUESTED** [review]

File: src/lib/assets/rbac.test.ts
Comment (samayer12) [inline src/lib/assets/rbac.test.ts:231]: These `CapabilityExport[]` objects seems to be re-used by a couple of tests. To be more direct about which keys were toggling for particular test cases, maybe we should have some sort of generator or a `defaultCapabilityExport` that we use as a starting point. Then, in each test case, the "Arrange" step involves setting the key(s) that are relevant to the particular test case.

The same idea goes for a lot of the "expected" JSON data

File: src/lib/assets/rbac.ts
Comment (samayer12) [inline src/lib/assets/rbac.ts]: Lingering TODO comment. Either fix it or write up an issue in the backlog so we don't lose track

File: src/lib/assets/deploy.ts
Comment (schaeferka) [reply src/lib/assets/deploy.ts]: Updated to pass in assets.config

File: src/lib/assets/rbac.test.ts
Comment (schaeferka) [reply src/lib/assets/rbac.test.ts:231]: Made a separate issue to revisit using a generator or `defaultCapabilityExport`.

File: src/lib/assets/rbac.ts
Comment (schaeferka) [reply src/lib/assets/rbac.ts]: Removed TODO

File: src/lib/assets/deploy.ts
Comment (samayer12) [reply src/lib/assets/deploy.ts]: We could do either way, I was curious if there was something about splitting them up that made more sense at first

**Review by samayer12 — DISMISSED** [review]

File: src/lib/assets/deploy.ts
Comment (cmwylie19) [inline src/lib/assets/deploy.ts]:
```suggestion
import { V1PolicyRule as PolicyRule } from "@kubernetes/client-node";
```

File: src/lib/assets/rbac.ts
Comment (cmwylie19) [inline src/lib/assets/rbac.ts]:
```suggestion
import { V1PolicyRule as PolicyRule } from "@kubernetes/client-node";
```

File: src/lib/assets/rbac.test.ts
Comment (cmwylie19) [inline src/lib/assets/rbac.test.ts]:
```suggestion
import { V1PolicyRule as PolicyRule } from "@kubernetes/client-node";
```

**Review by cmwylie19 — CHANGES_REQUESTED** [review]
need just one second on this.

File: src/lib/types.ts
Comment (cmwylie19) [inline src/lib/types.ts]:
```suggestion
import { V1PolicyRule as PolicyRule } from "@kubernetes/client-node";
```

File: src/lib/module.ts
Comment (cmwylie19) [inline src/lib/module.ts]:
```suggestion
import { V1PolicyRule as PolicyRule } from "@kubernetes/client-node";
```

**Review by cmwylie19 — DISMISSED** [review]

**Comment by cmwylie19** [comment]
The runners are not refreshing the code, the complaints are due to a change in KFC that was needed to unblock CI but no longer exists in this code base.

**Pull request closed without merging**
